README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Travis-CI Build Status](https://travis-ci.org/RSPaW/RSSApkg.svg?branch=master)](https://travis-ci.org/RSPaW/RSSApkg) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/RSPaW/RSSApkg?branch=master&svg=true)](https://ci.appveyor.com/project/RSPaW/RSSApkg)

RSSApkg
-------

RSSApkg is designed as a suite of functions to aid in some common tasks involving satellite imagery and other spatial data used in the Remote Sensing and Spatial Analysis section at Parks and Wildlife (Western Australia).

Requirements
------------

This package must be installed from GitHub. You can use the following code to install it:

``` r
library(devtools)
install_github("RSPaW/RSSApkg")
library(RSSApkg)
```

The package has dependencies with dplyr, lubridate, raster, rgdal, sp and stringr. On installation you may be prompted to install additional packages.

Help Files
----------

All functions contained in RSSApkg have their own help files and there is also the [RSSApkg website](https://rspaw.github.io/RSSApkg/index.html) with help files and vignettes.


NEWS.md
# RSSApkg 1.0.0

* Initial release of RSSApkg.
* Reworking of `jpegR` to:
    + enable begining and end dates to target analysis period.
    + output a small text file with name of original shape file.
    + split original shape file and store in new folder named `site_vectors`.
    + clean code and improve speed.
* Reworking of `extractR` to:
    + restrict min and max of dates in csv output to those utilised.
    + clean code and improve speed.
* Access to internal helper functions now grouped as utility functions and 
includes `u_dateR`, `u_leapR` and `u_shpsplitR`.





R/jpegR.R
#' Create small jpegs of satellite imagery for analysis.
#'
#' \code{jpegR} makes small jpeg images centered on and showing site locations
#' from a shape file. This function can be used as part of the cloud quality
#' check process prior to extracting satellite bands or indices. Although
#' designed to help the QA process the jpegs can be useful for quickly checking
#' many image dates for change (e.g. fires, vegetation removal, etc) or for
#' creating stills for animating. When run the function:
#' \enumerate{
#'     \item Obtains a list of imagery folders to access. In doing so it also
#'     checks for folder name errors (\code{\link{u_leapR}}) and only returns
#'     sun-corrected image folders.
#'     \item Splits a supplied shape file by unique attribute field entries,
#'     stores results in a folder and writes a text file containing the original
#'     shape file name (\code{\link{u_shpsplitR}}).
#'     \item Creates a small jpeg of the individual sites for each image date as
#'     per step 1. jpegs are based on band combination from user input and
#'     buffered by a desired distance.
#'     \item All outputs are placed in appropriately named folders per site in
#'     the working directory
#'     }
#'
#' @param wdir Character string filepath to the working directory where the
#' original shape file is located.
#' @param imdir Character string filepath to the path/row level for imagery.
#' @param layer Character string of the name of the shape file (no extension).
#' @param attrb Character string of the name of the field in the attribute
#' table of the shape file that contains the unique entries (e.g.site numbers).
#' @param start Character string representing start date of imagery search. Must
#' be in the form "dmY" or "dmy" and can include separators. Defaults to NA so
#' image search starts from first available image in the RSSA archive.
#' @param stop Character string representing end date of imagery search. Must
#' be in the form "dmY" or "dmy" and can include separators. Defaults to NA so
#' image search ends with latest available image in the RSSA archive.
#' @param combo Numerics representing bands to place in RGB channels of the jpeg.
#' Must be of the form c(N\emph{r}, N\emph{g}, N\emph{b}) where N\emph{r} is the
#' band number for the red channel etc.
#' @param buffer Numeric representing how much to buffer out the site to control
#' zoom level for jpegs. If imagery crs is projected, number represents metres.
#' If imagery is geodetic then the number represents degrees.
#'
#' @return Creates named folders, based on site location and dates of images
#' processed, containing small jpeg images for visual assessment.
#'
#' @author Bart Huntley, \email{bart.huntley@@dpaw.wa.gov.au}
#'
#' For more details see  \url{https://rspaw.github.io/RSSApkg/index.html}
#' {the RSSApkg website}
#'
#' @examples
#' \dontrun{
#' jpegR(wdir = "z:/working", imdir = "W:/usgs/115078", layer = "site_locations",
#' attrb = "ID", start = "1/1/98", stop = "28-12-2009", combo = c(5,4,2),
#' buffer = 2000)
#' }
#'
#' @export
#' @import raster
#' @importFrom rgdal readOGR
#' @importFrom lubridate dmy
#' @importFrom sp spTransform
#' @importFrom grDevices dev.off jpeg
jpegR <- function(wdir, imdir, layer, attrb, start=NA, stop=NA, combo,
                  buffer = 2000){
  #split layer, get names
  shpnames <- u_shpsplitR(pathin = wdir, layer = layer, attrb = attrb)

  #get folder/date imagery info
  alldo <- u_dateR(path = imdir, archive = TRUE)
  suppressWarnings(start <- lubridate::dmy(start))
  suppressWarnings(stop <- lubridate::dmy(stop))
  todo <- if (!is.na(start) & !is.na(stop)){
    sub <- subset(alldo, dates >= start & dates <= stop)#start and stop
    sub
  } else if (is.na(start) & !is.na(stop)){
    sub <- subset(alldo, dates <= stop)#stop only
    sub
  } else if (!is.na(start) & is.na(stop)){
    sub <- subset(alldo, dates >= start)#start only
    sub
  } else {
    alldo
  }

  #grab CRS of rasters for transform of layers
  proj <- raster::crs(raster::raster(todo[1, 1]))

  #make the jpegs
  for (i in seq_along(shpnames)){
    shp <- rgdal::readOGR(dsn = paste0(wdir, "/site_vectors"), shpnames[i])
    shp_t <- sp::spTransform(shp, proj)
    ext <- raster::extent(shp_t) + buffer
    beg <- todo[1, 2]
    end <- todo[length(todo[, 2]), 2]
    folder <- paste0(wdir, "/jpegs_site_", shpnames[i], "_", beg, "-", end)
    if (!file.exists(folder)){
      dir.create(folder)
      }
    for (j in seq_along(todo[, 1])){
      date <- todo[j, "dates"]
      jname <- paste0(date, "-", paste(combo, collapse = ""), ".jpg")
      fname <- paste0(folder, "/", jname)
      img <- todo[j, "path"]
      rstack <- raster::stack(img)
      jpeg(filename = fname, width = 842, height = 870)
      raster::plotRGB(rstack, r = combo[1], g = combo[2], combo[3],
                      ext = ext, stretch = "lin")
      raster::plot(shp_t, add = TRUE, lwd = 2, border = "magenta")
      dev.off()
    }

  }

}


R/extractR.R
#' Extracts mean values from Landsat imagery based on locations provided
#' by shape files.
#'
#'  \code{extractR} will extract mean pixel values from dates, determined from
#'  quality assessed jpegs, of Landsat raster imagery, from locations as
#'  provided by shape files. Common usage is for extracting vegetation indices
#'  from field work or calibration sites. When run it will:
#'   \enumerate{
#'       \item Copy the appropriate shape files to the appropriate quality
#'       assessed jpeg folders based on attribute id.
#'       \item Looks in each jpeg folder to determine appropriate image dates
#'       i.e. stores the dates of the remaining jpegs after quality assessment.
#'       \item Based on user input, extracts mean band or vegetation index
#'       value per site from the shape file.
#'       \item Writes mean values to a .csv file and stores each in appropriate
#'       jpeg folder.
#'       \item On completion it will join all extracted value data sets into one
#'       .csv file.
#'       }
#'
#' @param wdir Character string filepath to the working directory where the
#' original shape files are located. This working directory must contain a
#' folder named "site_vectors" which contains the individual shape files for
#' extraction.
#' @param imdir Character string filepath to the path/row level for imagery.
#' @param option A character string indicating one of the following "i35",
#' "ndvi", "b1", "b2", "b3", "b4", "b5", "b6".
#' @param attrb A character string of the name of field in the attribute column
#' of the shape file that contains the unique identifier for the location.
#' This must be the same as was used to create the jpegs and is usually the site
#' ID.
#'
#' @return Creates individual .csv files for extracted values per shape file and
#' another that collates them all.
#'
#'
#' @author Bart Huntley, \email{bart.huntley@dpaw.wa.gov.au}
#'
#' For more details see  \url{https://rspaw.github.io/RSSApkg/index.html}
#' {the RSSApkg website}
#'
#' @examples
#' \dontrun{
#' extractR(wdir = "z:/working", imdir = "W:/usgs/115078", option = "i35", attrb = "ID")
#' }
#'
#' @export
#' @importFrom stringr str_replace_all
#' @importFrom rgdal readOGR
#' @importFrom raster crs raster extract
#' @importFrom sp spTransform
#' @importFrom lubridate ymd
#' @importFrom dplyr left_join
#' @importFrom utils write.csv write.table
extractR <- function(wdir, imdir, option, attrb){
  #function for right substring
  substrRight <- function(x, n){
    substr(x, nchar(x) - n + 1, nchar(x))
  }
  #get jpeg folders
  jfolds <- list.files(path = wdir, pattern = "jpegs_site")

  #get vector folder
  vfold <- paste0(wdir, "/site_vectors")

  #copy shape files to relevant QA folders
  shpfiles <- list.files(path = vfold, pattern = "*.shp$")
  shpnames <- unlist(strsplit(shpfiles, split = "\\.")) [c(TRUE, FALSE)]
  for (i in seq_along(shpnames)){
    shp <- shpnames[i]
    patt1 <- paste0("^", shp, "\\.")
    shps <- list.files(path = vfold, pattern = patt1)
    from <- paste0(vfold, "/", shps)
    patt2 <- paste0("site_", shp, "_")
    to <- paste0(wdir, "/", jfolds[grep(patt2, jfolds)])
    file.copy(from = from, to = to, recursive = FALSE, overwrite = TRUE,
              copy.mode = TRUE)
  }

  #get jpeg dates
  jlist <- vector("list", length(jfolds))
  names(jlist) <- jfolds

  for (j in seq_along(jfolds)){
    jpegs <- list.files(path = paste0(wdir, "/", jfolds[j]), pattern = "*.jpg")
    jdates <- substr(jpegs, 1, 10)
    ifolds <- stringr::str_replace_all(jdates, "[^[:alnum:]]", "")
    jlist[[j]] <- paste0(imdir, "/", ifolds)
  }

  #empty list for storage of all csvs at end
  resultslist <- vector("list", length(jfolds))

  #extract to csv
  for (k in seq_along(jlist)){
    #get site shape file
    shpk <- shpnames[k]
    sitesSHP <- rgdal::readOGR(dsn = paste0(wdir, "/", names(jlist[k])), shpk)
    rnames <- as.character(sitesSHP@data[, attrb])

    #get beginning and end dates
    beg <- substrRight(jlist[[k]][1], 8)
    end <- substrRight(jlist[[k]][length(jlist[[k]])], 8)

    #make sure same projections (shape to raster)
    rfile <- list.files(path = jlist[[1]][1], pattern = "pre.ers",
                        full.names = TRUE)[1]
    rastproj <- raster::crs(raster::raster(rfile))
    sitesSHPt <- sp::spTransform(sitesSHP, rastproj)

    #empty df for results and options vector
    results <- as.data.frame(matrix(ncol = length(rnames) + 1,
                                    nrow = length(jlist[k][[1]]),
                                    dimnames = list(NULL, c("date", rnames))),
                             stringsAsFactors = FALSE)


    #extract from imagery rasters
    for (m in seq_along(jlist[k][[1]])){

      #image file name
      imfile <- list.files(path = jlist[k][[1]][m], pattern = "pre.ers",
                           full.names = TRUE)

      #make sure only one pre and return complete path
      if (length(imfile) > 1){
        imdat <- imfile[grepl("USG", imfile)]
      } else {
        imdat <- imfile
      }

      #grab path row and date
      prow <- substr(imdat, 9, 14)
      date <- lubridate::ymd(substr(imdat, 16, 23))

      #extract by option
      if (option == "i35"){
        b3r <- raster::raster(imdat, band = 3)
        b5r <- raster::raster(imdat, band = 5)
        b3x <- raster::extract(b3r, sitesSHPt, fun = mean, na.rm = TRUE)
        b5x <- raster::extract(b5r, sitesSHPt, fun = mean, na.rm = TRUE)
        i35 <- (b3x + b5x) / 2
        out <- i35[1, 1]
      } else if (option == "ndvi"){
        b3r <- raster::raster(imdat, band = 3)
        b4r <- raster::raster(imdat, band = 4)
        b3x <- raster::extract(b3r, sitesSHPt, fun = mean, na.rm = TRUE)
        b4x <- raster::extract(b4r, sitesSHPt, fun = mean, na.rm = TRUE)
        ndvi <- (b4x - b3x) / (b4x + b3x)
        out <- ndvi[1, 1]
      } else if (option == "b1"){
        b1r <- raster::raster(imdat, band = 1)
        b1x <- raster::extract(b1r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b1x[1, 1]
      } else if (option == "b2"){
        b2r <- raster::raster(imdat, band = 2)
        b2x <- raster::extract(b2r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b2x[1, 1]
      } else if (option == "b3"){
        b3r <- raster::raster(imdat, band = 3)
        b3x <- raster::extract(b3r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b3x[1, 1]
      } else if (option == "b4"){
        b4r <- raster::raster(imdat, band = 4)
        b4x <- raster::extract(b4r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b4x[1, 1]
      } else if (option == "b5"){
        b5r <- raster::raster(imdat, band = 5)
        b5x <- raster::extract(b5r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b5x[1, 1]
      } else {
        b6r <- raster::raster(imdat, band = 6)
        b6x <- raster::extract(b6r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b6x[1, 1]
      }

      results[m, 1] <- as.character(date)
      results[m, 2] <- out

    }

    ## write out results per shp file
    write.csv(file = paste0(wdir, "/", jfolds[k], "/", prow, "_", option, "_",
                            shpk, "_", beg, "-", end, ".csv"), x = results,
              row.names = FALSE)
    resultslist[[k]] <- results

  }

  #combine all data into 1 csv
  #get all poss image dates and subset to match data range
  first_elems <- unlist(lapply(jlist, head, n = 1L))
  first_date <- lubridate::ymd(sort(substrRight(first_elems, 8))[1])
  last_elems <- unlist(lapply(jlist, tail, n = 1L))
  last_date <- lubridate::ymd(sort(substrRight(last_elems, 8),
                                   decreasing = TRUE))[1]


  allfolds <- suppressWarnings(lubridate::ymd(list.files(path = imdir)))
  alldates <- allfolds[!is.na(allfolds)]
  subdates <- alldates[alldates >= first_date & alldates <= last_date]
  alldat <- data.frame(date = as.character(subdates), stringsAsFactors = FALSE)


  for (n in seq_along(resultslist)){
    alldat <- dplyr::left_join(alldat, resultslist[[n]], "date")
  }
  write.csv(x = alldat, file = paste0(wdir, "/", prow, "_", option, "_QA_", beg,
                                      "-", end, ".csv"))
}


R/RSSA_utils.R
# utility helper functions

#' Renames incorrectly named leap year image date folders.
#'
#' \code{u_leapR} reads the date of a suncorrected image, checks it against the
#' folder name and if incorrect, renames the folder. This corrects for leap year
#' folder naming errors propagated by batch processing.
#'
#' @param path Character string filepath to the path/row of imagery to check.
#'
#' @return If incorrectly named folders are identified they are renamed a day
#'     earlier, otherwise a message "No leap date folder errors" is printed to
#'     screen.
#'
#' @author Bart Huntley, \email{bart.huntley@@dpaw.wa.gov.au}
#'
#' @examples
#' \dontrun{
#' u_leapR("W:/usgs/115078")
#' }
#'
#' @export
#' @importFrom lubridate ymd dmy

u_leapR <- function(path){
  allfiles <- list.files(path = path, recursive = TRUE)
  result <- allfiles[grepl("*pre.ers", allfiles)]
  result <- result[!grepl("^[a-zA-Z]", result)]
  #date for folder
  fold <- substr(result, 1, 8)
  fdate <- as.character(lubridate::ymd(fold))
  #date for image
  ldate <- as.character(lubridate::dmy(substr(result, 21, 26)))
  #find mismatch
  bad_fold_dates <- setdiff(fdate, ldate)
  if (length(bad_fold_dates) > 0){
    #correct the folder names and path
    corr_fold_dates <- lubridate::ymd(bad_fold_dates) - 1
    corr_fold <- gsub("-", "", as.character(corr_fold_dates))
    new_name <- paste0(path, "\\", corr_fold)
    #old folder names and path to correct
    bad_fold <- gsub("-", "", as.character(bad_fold_dates))
    old_name <- paste0(path, "\\", bad_fold)
    #rename folders
    file.rename(old_name, new_name)
  } else {
    cat("No leap date folder errors\n")
  }

}

#' Creates a data frame of paths, folder names and dates.
#'
#' \code{u_dateR} creates a data frame of paths, folder names and dates from the
#' contents of a folder.
#'
#' This output of this function will hold paths, folder names and dates that will
#' be used for other functions such as \code{jpegR} and \code{extractR}.
#'
#' @param path Character string filepath to the path/row of satellite imagery to
#'     check. Can also be a folder of jpegs or some other date named files.
#'
#' @param archive A logical scalar. \strong{TRUE}  will assume \strong{path} is
#'     indicating RSSA internal USGS archive. \strong{FALSE} can be used when
#'     querying a QA jpeg folder or the like.
#'
#' @param pat Pattern to search for in a non-USGS imagery archive. Defaults to
#'     ".jpeg".
#'
#' @return For a USGS imagery archive the query will return a data frame with:
#' \itemize{
#'     \item path - a column of character string file paths
#'     \item folds - a column of character string folder names
#'     \item dates - a column of date class dates
#'     }
#' For a non-USGS archive the query will return a data frame with:
#' \itemize{
#'     \item path - a column of character string file paths
#'     \item folds - a column of character string folder names
#'     }
#' @examples
#' \dontrun{
#' u_dateR(path = "W:/usgs/115078", archive = TRUE)
#' u_dateR(path = "Z:/DEC/working/QA...", archive = FALSE, pat = ".jpeg")
#' }
#'
#' @author Bart Huntley, \email{bart.huntley@@dpaw.wa.gov.au}
#'
#' @export
#' @importFrom lubridate ymd

u_dateR <- function(path, archive, pat = ".jpeg"){
  if (archive == TRUE){
    u_leapR(path)
    prefolds <- list.files(path = path, pattern = "*pre.ers$", recursive = TRUE)
    ind <- grepl("^[[:digit:]]", prefolds)
    prefolds <- prefolds[ind]
    datesdf <- data.frame(path = paste0(path, "/", prefolds),
                          folds = substr(prefolds, 1, 8),
                          dates = lubridate::ymd(substr(prefolds, 1, 8)),
                          stringsAsFactors = FALSE)
    return(datesdf)
  } else {
    files <- list.files(path = path, pattern = pat)
    datesdf <- data.frame(folds = gsub("-", "", substr(files, 1, 10)),
                          dates = lubridate::ymd(substr(files, 1, 10)),
                          stringsAsFactors = FALSE)
    return(datesdf)
  }
}

#' A tool for splitting ESRI shape files.
#'
#' \code{u_shpsplitR} takes an ESRI shape file and splits it into multiple shape
#' files based on unique entries in a field of the attribute table.
#'
#' @param pathin Character string filepath to the location of the shape file.
#' @param pathout Character string filepath to the write location.
#' @param layer Character string of the name of the shape file (no extension).
#' @param attrb Character string of the name of the field in the attribute
#' table of the shape file that contains the unique entries (e.g.site numbers).
#'
#' @return Creates multiple ESRI shape files each named after the unique entry
#' found in 'attrb' and saves them to a folder (defaults to "site_vectors"). It
#' will also write a text file with the name of the original layer file used and
#' output names of new layers created to screen or to object if created.
#'
#' @author Bart Huntley, \email{bart.huntley@@dpaw.wa.gov.au}
#'
#' @examples
#' \dontrun{
#' u_shpsplitR(layer = "plot_locations", attrb = "Plot_ID")
#' }
#'
#' @export
#' @importFrom  rgdal readOGR writeOGR

u_shpsplitR <- function(pathin, pathout = paste0(pathin, "/site_vectors"),
                        layer, attrb){
  if (!file.exists("site_vectors")){
    dir.create("site_vectors")
    }
  data <-  rgdal::readOGR(dsn = pathin, layer)
  sites <- as.character(unique(data@data[, attrb]))
  for (i in 1:length(sites)){
    tmp <- data[data@data[, attrb] == sites[i], ]
    rgdal::writeOGR(tmp, dsn = pathout, sites[i], driver = "ESRI Shapefile",
                    overwrite_layer = TRUE)
    write.table(layer, paste0(pathout, "/OriginShapeFile.txt"),
                row.names = FALSE, quote = FALSE, col.names = FALSE)
  }
  return(sites)
}


working.R
# utility functions

# function to get dates
# u_dates

path <- "W:/usgs/115078"
archive <- TRUE

path <- paste0("Z:/DOCUMENTATION/BART/R/R_DEV/devtest")
archive2 <- FALSE
pat <- ".jpeg"

setwd(path)

u_leapR <- function(path){
  allfiles <- list.files(path = path, recursive = TRUE)
  result <- allfiles[grepl("*pre.ers", allfiles)]
  result <- result[!grepl("^[a-zA-Z]", result)]
  #date for folder
  fold <- substr(result, 1, 8)
  fdate <- as.character(lubridate::ymd(fold))
  #date for image
  ldate <- as.character(lubridate::dmy(substr(result, 21, 26)))
  #find mismatch
  bad.fold.dates <- setdiff(fdate, ldate)
  if(length(bad.fold.dates) > 0){
    #correct the folder names and path
    corr.fold.dates <- lubridate::ymd(bad.fold.dates) - 1
    corr.fold <- gsub("-", "", as.character(corr.fold.dates))
    new.name <- paste0(path, "\\", corr.fold)
    #old folder names and path to correct
    bad.fold <- gsub("-", "", as.character(bad.fold.dates))
    old.name <- paste0(path, "\\", bad.fold)
    #rename folders
    file.rename(old.name, new.name)
  } else {
    cat("No leap date folder errors")
  }

}
u_dateR <- function(path, archive, pat = ".jpeg"){
  if(archive == TRUE){
    u_leapR(path)
    prefolds <- list.files(path = path, pattern = "*pre.ers$", recursive = TRUE)
    ind <- grepl("^[[:digit:]]", prefolds)
    prefolds <- prefolds[ind]
    datesdf <- data.frame(path = paste0(path, "/", prefolds),
                          folds = substr(prefolds, 1, 8),
                          dates = lubridate::ymd(substr(prefolds, 1, 8)),
                          stringsAsFactors = FALSE)
    return(datesdf)
  } else {
    #get jpeg dirs
    dirs <- list.dirs()
    dirs <- dirs[grepl("/jpegs_", dirs)]

    #create list
    qadlist <- vector("list", length(dirs))
    names(qadlist) <- dirs

    for(h in 1:length(dirs)){
      jpegs <- list.files(path = dirs[h], pattern = "*.jpeg")
      qadates <- substr(jpegs, 1, 10)
      qafolds <- stringr::str_replace_all(qadates, "[^[:alnum:]]", "")
      qadlist[[h]] <- paste0("holder", "/", qafolds)
    }
    # files <- list.files(path = path, pattern = pat)
    # datesdf <- data.frame(folds = gsub("-", "", substr(files, 1, 10)),
    #                       dates = lubridate::ymd(substr(files, 1, 10)),
    #                       stringsAsFactors = FALSE)
    return(qadlist)
  }
}
u_shpsplitR <- function(pathin = ".", pathout = "./site_vectors", layer, attrb){
  if(!file.exists("site_vectors")){ dir.create("site_vectors")}
  data <-  rgdal::readOGR(dsn = pathin, layer)
  sites <- as.character(unique(data@data[,attrb]))
  for(i in 1:length(sites)){
    tmp <- data[data@data[,attrb] == sites[i], ]
    rgdal::writeOGR(tmp, dsn = pathout, sites[i], driver = "ESRI Shapefile",
                    overwrite_layer = TRUE)
    write.table(layer, paste0(pathout, "/OriginShapeFile.txt"),
                row.names = FALSE, quote = FALSE, col.names = FALSE)
  }
  return(sites)
}
test <- u_dateR(path, archive)
test2 <- u_dateR(path2, archive2)

wdir <- "Z:/DOCUMENTATION/BART/R/R_DEV/devtest"
imdir <- "W:/usgs/108083"
#setwd(wdir)
layer <- "SWWMP_wetlands_10883_20170227"
attrb <- "WCode"
start <- "1/1/2000"
stop <- "01/06/2000"
combo <- c(5,4,2)
#buffer <- 2000

jpegR <- function(wdir=".", imdir, layer, attrb, start=NA, stop=NA, combo,
                  buffer = 2000){
  #split layer, get names
  shpnames <- u_shpsplitR(layer = layer, attrb = attrb)

  #get folder/date imagery info
  alldo <- u_dateR(path = imdir, archive = TRUE)
  suppressWarnings(start <- lubridate::dmy(start))
  suppressWarnings(stop <- lubridate::dmy(stop))
  todo <- if(!is.na(start) & !is.na(stop)){
    sub <- subset(alldo, dates >= start & dates <= stop)#start and stop
    sub
  } else if(is.na(start) & !is.na(stop)){
    sub <- subset(alldo, dates <= stop)#stop only
    sub
  } else if(!is.na(start) & is.na(stop)){
    sub <- subset(alldo, dates >= start)#start only
    sub
  } else {
    alldo
  }

  #grab CRS of rasters for transform of layers
  proj <- raster::crs(raster::raster(todo[1,1]))

  #make the jpegs
  for(i in seq_along(shpnames)){
    shp <- rgdal::readOGR(dsn = "./site_vectors", shpnames[i])
    shp_t <- sp::spTransform(shp, proj)
    ext <- raster::extent(shp_t) + buffer
    beg <- todo[1,2]
    end <- todo[length(todo[,2]), 2]
    folder <- paste0("jpegs_site_", shpnames[i], "_", beg, "-", end)
    if(!file.exists(folder)){ dir.create(folder)}
    for(j in seq_along(todo[,1])){
      date <- todo[j,"dates"]
      jname <- paste0(date, "-", paste(combo, collapse=""), ".jpeg")
      fname <- paste0("./", folder, "/", jname)
      img <- todo[j,"path"]
      rstack <- raster::stack(img)
      jpeg(filename = fname, width = 842, height = 870)
      raster::plotRGB(rstack, r = combo[1], g = combo[2], combo[3],
                      ext = ext, stretch = "lin")
      raster::plot(shp_t, add = TRUE, lwd = 2, border = "magenta")
      dev.off()
    }

  }

}

jpegR(wdir, imdir, layer, attrb, start, stop, combo, buffer = 2000)

##extract new
wdir <- "Z:/DOCUMENTATION/BART/R/R_DEV/devtest"
imdir <- "W:/usgs/108083"
option <- "i35"
attrb <- "WCode"
setwd(wdir)
extractR(wdir, imdir, option, attrb)

extractR <- function(wdir, imdir, option, attrb){
  #function for right substring
  substrRight <- function(x, n){
    substr(x, nchar(x)-n+1, nchar(x))
  }
  #get jpeg folders
  jfolds <- list.files(path = wdir, pattern = "jpegs_site")

  #get vector folder
  vfold <- "./site_vectors"

  #copy shape files to relevant QA folders
  shpfiles <- list.files(path = vfold, pattern = "*.shp$")
  shpnames <- unlist(strsplit(shpfiles, split = "\\."))[c(TRUE,FALSE)]
  for(i in seq_along(shpnames)){
    shp <- shpnames[i]
    patt1 <- paste0("^", shp, "\\.")
    shps <- list.files(path = vfold, pattern = patt1)
    from <- paste0(vfold,"/", shps)
    patt2 <- paste0("site_", shp, "_")
    to <- paste0(wdir, "/", jfolds[grep(patt2, jfolds)])
    file.copy(from=from, to=to, recursive = FALSE, overwrite = TRUE,
              copy.mode = TRUE)
  }

  #get jpeg dates
  jlist <- vector("list", length(jfolds))
  names(jlist) <- jfolds

  for(j in seq_along(jfolds)){
    jpegs <- list.files(path = paste0("./", jfolds[j]), pattern = "*.jpeg")
    jdates <- substr(jpegs, 1, 10)
    ifolds <- stringr::str_replace_all(jdates, "[^[:alnum:]]", "")
    jlist[[j]] <- paste0(imdir, "/", ifolds)
  }

  #empty list for storage of all csvs at end
  resultslist <- vector("list", length(jfolds))

  #extract to csv
  for(k in seq_along(jlist)){
    #get site shape file
    shpk <- shpnames[k]
    sitesSHP <- rgdal::readOGR(dsn = paste0("./", names(jlist[k])), shpk)
    rnames <- as.character(sitesSHP@data[, attrb])

    #get beginning and end dates
    beg <- substrRight(jlist[[k]][1], 8)
    end <- substrRight(jlist[[k]][length(jlist[[k]])], 8)

    #make sure same projections (shape to raster)
    rfile <- list.files(path = jlist[[1]][1], pattern = "pre.ers", full.names = TRUE)[1]
    rastproj <- raster::crs(raster::raster(rfile))
    sitesSHPt <- sp::spTransform(sitesSHP, rastproj)

    #empty df for results and options vector
    results <- as.data.frame(matrix(ncol=length(rnames) + 1,
                                    nrow = length(jlist[k][[1]]),
                                    dimnames=list(NULL, c("date", rnames))),
                             stringsAsFactors = FALSE)
    options <- c("i35", "ndvi", "b1", "b2", "b3", "b4", "b5", "b6")


    #extract from imagery rasters
    for(m in seq_along(jlist[k][[1]])){

      #image file name
      imfile <- list.files(path = jlist[k][[1]][m], pattern = "pre.ers",
                           full.names = TRUE)

      #make sure only one pre and return complete path
      if(length(imfile) > 1){
        imdat <- imfile[grepl("USG", imfile)]
      } else { imdat <- imfile}

      #grab path row and date
      prow <- substr(imdat, 9, 14)
      date <- lubridate::ymd(substr(imdat, 16, 23))

      #extract by option
      if(option == "i35"){
        b3r <- raster::raster(imdat, band = 3)
        b5r <- raster::raster(imdat, band = 5)
        b3x <- raster::extract(b3r, sitesSHPt, fun = mean, na.rm = TRUE)
        b5x <- raster::extract(b5r, sitesSHPt, fun = mean, na.rm = TRUE)
        i35 <- (b3x + b5x)/2
        out <- i35[1,1]
      } else if (option == "ndvi"){
        b3r <- raster::raster(imdat, band = 3)
        b4r <- raster::raster(imdat, band = 4)
        b3x <- raster::extract(b3r, sitesSHPt, fun = mean, na.rm = TRUE)
        b4x <- raster::extract(b4r, sitesSHPt, fun = mean, na.rm = TRUE)
        ndvi <- (b4x - b3x)/(b4x + b3x)
        out <- ndvi[1,1]
      } else if (option == "b1"){
        b1r <- raster::raster(imdat, band = 1)
        b1x <- raster::extract(b1r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b1x[1,1]
      } else if (option == "b2"){
        b2r <- raster::raster(imdat, band = 2)
        b2x <- raster::extract(b2r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b2x[1,1]
      } else if (option == "b3"){
        b3r <- raster::raster(imdat, band = 3)
        b3x <- raster::extract(b3r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b3x[1,1]
      } else if (option == "b4"){
        b4r <- raster::raster(imdat, band = 4)
        b4x <- raster::extract(b4r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b4x[1,1]
      } else if (option == "b5"){
        b5r <- raster::raster(imdat, band = 5)
        b5x <- raster::extract(b5r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b5x[1,1]
      } else {
        b6r <- raster::raster(imdat, band = 6)
        b6x <- raster::extract(b6r, sitesSHPt, fun = mean, na.rm = TRUE)
        out <- b6x[1,1]
      }

      results[m,1] <- as.character(date)
      results[m,2] <- out

    }

    ## write out results per shp file
    write.csv(file = paste0("./", jfolds[k], "/", prow,"_", option, "_", shpk,
                            "_", beg, "-", end, ".csv"),
              x = results, row.names = FALSE)
    resultslist[[k]] <- results

  }
  ## Combine all results in one data source ####################################

  #combine all data into 1 csv
  #get all poss image dates and subset to match data range
  first_elems <- unlist(lapply(jlist, head, n = 1L))
  first_date <- lubridate::ymd(sort(substrRight(first_elems, 8))[1])
  last_elems <- unlist(lapply(jlist, tail, n = 1L))
  last_date <- lubridate::ymd(sort(substrRight(last_elems, 8),
                                         decreasing = TRUE))[1]


  allfolds <- suppressWarnings(lubridate::ymd(list.files(path = imdir)))
  alldates <- allfolds[!is.na(allfolds)]
  subdates <- alldates[alldates >= first_date & alldates <= last_date]
  alldat <- data.frame(date = as.character(subdates), stringsAsFactors = FALSE)


  for(n in seq_along(resultslist)){
    alldat <- dplyr::left_join(alldat, resultslist[[n]], "date")
  }
  write.csv(x = alldat, file = paste0(prow, "_", option, "_QA_", beg, "-", end,
                                  ".csv"))
}


